File: bindings/csharp/examples/BasicExample.cs
using System;
using System.Net.Http;
using System.Threading.Tasks;
using MetaOxide;

namespace MetaOxide.Examples
{
    /// <summary>
    /// Basic example demonstrating metadata extraction with MetaOxide.
    /// </summary>
    public class BasicExample
    {
        public static async Task Main(string[] args)
        {
            Console.WriteLine("MetaOxide Basic Example");
            Console.WriteLine("=======================\n");

            // Example 1: Extract from local HTML string
            Console.WriteLine("1. Extract from local HTML:");
            ExtractFromString();

            Console.WriteLine("\n" + new string('-', 60) + "\n");

            // Example 2: Extract from URL
            Console.WriteLine("2. Extract from URL:");
            await ExtractFromUrl("https://example.com");

            Console.WriteLine("\n" + new string('-', 60) + "\n");

            // Example 3: Extract specific formats
            Console.WriteLine("3. Extract specific formats:");
            ExtractSpecificFormats();
        }

        private static void ExtractFromString()
        {
            var html = @"
<!DOCTYPE html>
<html>
<head>
    <title>Example Page</title>
    <meta name=""description"" content=""This is an example page demonstrating MetaOxide"">
    <meta name=""keywords"" content=""example,metadata,extraction"">
    <meta property=""og:title"" content=""Example Page - Open Graph"">
    <meta property=""og:type"" content=""website"">
    <meta property=""og:url"" content=""https://example.com"">
    <meta name=""twitter:card"" content=""summary_large_image"">
    <meta name=""twitter:title"" content=""Example Page - Twitter"">
</head>
<body>
    <h1>Welcome</h1>
</body>
</html>";

            try
            {
                // Extract all metadata
                var result = Extractor.ExtractAll(html, "https://example.com");

                Console.WriteLine($"Found {result.GetMetadataFormatCount()} metadata formats\n");

                // Display meta tags
                if (result.Meta != null)
                {
                    Console.WriteLine("Standard Meta Tags:");
                    foreach (var kvp in result.Meta)
                    {
                        Console.WriteLine($"  {kvp.Key}: {kvp.Value}");
                    }
                }

                // Display Open Graph
                if (result.OpenGraph != null)
                {
                    Console.WriteLine("\nOpen Graph:");
                    foreach (var kvp in result.OpenGraph)
                    {
                        Console.WriteLine($"  {kvp.Key}: {kvp.Value}");
                    }
                }

                // Display Twitter Card
                if (result.Twitter != null)
                {
                    Console.WriteLine("\nTwitter Card:");
                    foreach (var kvp in result.Twitter)
                    {
                        Console.WriteLine($"  {kvp.Key}: {kvp.Value}");
                    }
                }
            }
            catch (MetaOxideException ex)
            {
                Console.WriteLine($"Error: {ex.GetFriendlyMessage()}");
            }
        }

        private static async Task ExtractFromUrl(string url)
        {
            using var httpClient = new HttpClient();
            httpClient.DefaultRequestHeaders.UserAgent.ParseAdd("MetaOxide Example/1.0");

            try
            {
                // Fetch HTML
                var html = await httpClient.GetStringAsync(url);

                // Extract metadata
                var result = Extractor.ExtractAll(html, url);

                Console.WriteLine($"Extracted metadata from: {url}");
                Console.WriteLine($"Found {result.GetMetadataFormatCount()} metadata formats");

                // Display summary
                if (result.Meta != null && result.Meta.ContainsKey("description"))
                {
                    Console.WriteLine($"\nDescription: {result.Meta["description"]}");
                }

                if (result.OpenGraph != null && result.OpenGraph.ContainsKey("title"))
                {
                    Console.WriteLine($"OG Title: {result.OpenGraph["title"]}");
                }

                // Export to JSON
                var json = result.ToJson();
                Console.WriteLine($"\nJSON output ({json.Length} characters)");
            }
            catch (HttpRequestException ex)
            {
                Console.WriteLine($"HTTP Error: {ex.Message}");
            }
            catch (MetaOxideException ex)
            {
                Console.WriteLine($"Extraction Error: {ex.GetFriendlyMessage()}");
            }
        }

        private static void ExtractSpecificFormats()
        {
            var html = @"
<!DOCTYPE html>
<html>
<head>
    <meta property=""og:title"" content=""Open Graph Title"">
    <meta property=""og:description"" content=""OG Description"">
    <meta name=""twitter:card"" content=""summary"">
    <script type=""application/ld+json"">
    {
        ""@context"": ""https://schema.org"",
        ""@type"": ""Article"",
        ""headline"": ""Example Article"",
        ""author"": ""John Doe""
    }
    </script>
</head>
</html>";

            // Extract only Open Graph
            var openGraph = Extractor.ExtractOpenGraph(html);
            Console.WriteLine("Open Graph only:");
            if (openGraph != null)
            {
                foreach (var kvp in openGraph)
                {
                    Console.WriteLine($"  {kvp.Key}: {kvp.Value}");
                }
            }

            // Extract only Twitter Card
            var twitter = Extractor.ExtractTwitter(html);
            Console.WriteLine("\nTwitter Card only:");
            if (twitter != null)
            {
                foreach (var kvp in twitter)
                {
                    Console.WriteLine($"  {kvp.Key}: {kvp.Value}");
                }
            }

            // Extract only JSON-LD
            var jsonLd = Extractor.ExtractJsonLd(html);
            Console.WriteLine("\nJSON-LD only:");
            if (jsonLd != null)
            {
                Console.WriteLine($"  Found {jsonLd.Count} JSON-LD item(s)");
            }
        }
    }
}

File: bindings/java/meta-oxide-java/examples/BasicExample.java
package io.github.yfedoseev.metaoxide.examples;

import io.github.yfedoseev.metaoxide.Extractor;
import io.github.yfedoseev.metaoxide.ExtractionResult;
import io.github.yfedoseev.metaoxide.MetaOxideException;

import java.util.Map;

/**
 * Basic example demonstrating MetaOxide metadata extraction in Java.
 * <p>
 * This example shows how to:
 * - Extract all metadata formats at once
 * - Access individual metadata fields
 * - Handle errors properly
 * </p>
 */
public class BasicExample {

    public static void main(String[] args) {
        // Sample HTML with various metadata formats
        String html = "<!DOCTYPE html>\n" +
                "<html>\n" +
                "<head>\n" +
                "    <title>Amazing Article About Rust</title>\n" +
                "    <meta name=\"description\" content=\"Learn about Rust programming language\">\n" +
                "    <meta name=\"keywords\" content=\"rust, programming, systems\">\n" +
                "    <meta property=\"og:title\" content=\"Amazing Article About Rust\">\n" +
                "    <meta property=\"og:type\" content=\"article\">\n" +
                "    <meta property=\"og:url\" content=\"https://example.com/article\">\n" +
                "    <meta property=\"og:image\" content=\"https://example.com/image.jpg\">\n" +
                "    <meta name=\"twitter:card\" content=\"summary_large_image\">\n" +
                "    <meta name=\"twitter:title\" content=\"Amazing Article\">\n" +
                "    <script type=\"application/ld+json\">\n" +
                "    {\n" +
                "        \"@context\": \"https://schema.org\",\n" +
                "        \"@type\": \"Article\",\n" +
                "        \"headline\": \"Amazing Article About Rust\",\n" +
                "        \"author\": {\n" +
                "            \"@type\": \"Person\",\n" +
                "            \"name\": \"John Doe\"\n" +
                "        },\n" +
                "        \"datePublished\": \"2024-01-15\"\n" +
                "    }\n" +
                "    </script>\n" +
                "</head>\n" +
                "<body>\n" +
                "    <article class=\"h-entry\">\n" +
                "        <h1 class=\"p-name\">Getting Started with Rust</h1>\n" +
                "        <time class=\"dt-published\" datetime=\"2024-01-15\">January 15, 2024</time>\n" +
                "    </article>\n" +
                "</body>\n" +
                "</html>";

        try {
            // Extract ALL metadata at once (most efficient!)
            System.out.println("=== Extracting All Metadata ===\n");
            ExtractionResult result = Extractor.extractAll(html, "https://example.com");

            // Standard HTML meta tags
            System.out.println("Standard Meta Tags:");
            System.out.println("  Title: " + result.meta.get("title"));
            System.out.println("  Description: " + result.meta.get("description"));
            System.out.println("  Keywords: " + result.meta.get("keywords"));
            System.out.println();

            // Open Graph Protocol
            System.out.println("Open Graph:");
            System.out.println("  Title: " + result.openGraph.get("title"));
            System.out.println("  Type: " + result.openGraph.get("type"));
            System.out.println("  URL: " + result.openGraph.get("url"));
            System.out.println("  Image: " + result.openGraph.get("image"));
            System.out.println();

            // Twitter Cards
            System.out.println("Twitter Cards:");
            System.out.println("  Card: " + result.twitter.get("card"));
            System.out.println("  Title: " + result.twitter.get("title"));
            System.out.println();

            // JSON-LD
            System.out.println("JSON-LD:");
            if (!result.jsonLd.isEmpty()) {
                @SuppressWarnings("unchecked")
                Map<String, Object> article = (Map<String, Object>) result.jsonLd.get(0);
                System.out.println("  Type: " + article.get("@type"));
                System.out.println("  Headline: " + article.get("headline"));

                @SuppressWarnings("unchecked")
                Map<String, Object> author = (Map<String, Object>) article.get("author");
                if (author != null) {
                    System.out.println("  Author: " + author.get("name"));
                }
                System.out.println("  Published: " + article.get("datePublished"));
            }
            System.out.println();

            // Microformats
            System.out.println("Microformats:");
            if (result.microformats.containsKey("h-entry")) {
                System.out.println("  Found h-entry microformat");
            }
            System.out.println();

            // Summary
            System.out.println(result);
            System.out.println();

            // Extract individual formats (less efficient, but available)
            System.out.println("=== Extracting Individual Formats ===\n");

            Map<String, Object> meta = Extractor.extractMeta(html, "https://example.com");
            System.out.println("Meta tags count: " + meta.size());

            Map<String, Object> og = Extractor.extractOpenGraph(html, "https://example.com");
            System.out.println("Open Graph properties: " + og.size());

            // Library version
            System.out.println("\nMetaOxide version: " + Extractor.getVersion());

        } catch (MetaOxideException e) {
            System.err.println("Error extracting metadata: " + e.getMessage());
            System.err.println("Error code: " + e.getErrorCode());
            e.printStackTrace();
        }
    }
}
